docs/Secure-Coding-Guide-for-Python/CWE-682/CWE-191/compliant02.py

```diff
@@ -1,23 +1,71 @@
 # SPDX-FileCopyrightText: OpenSSF project contributors
 # SPDX-License-Identifier: MIT
-""" Compliant Code Example """
+"""Compliant Code Example"""
 
-import time
+from datetime import datetime, timedelta
+import logging
 
+# Enabling verbose debugging
+# logging.basicConfig(level=logging.DEBUG)
 
-def get_time_in_future(hours_in_future):
-    """Gets the time n hours in the future"""
-    try:
-        currtime = list(time.localtime())
-        currtime[3] = currtime[3] + hours_in_future
-        if currtime[3] + hours_in_future > 24:
-            currtime[3] = currtime[3] - 24
-        return time.asctime(tuple(currtime)).split(" ")[3]
-    except OverflowError as _:
-        return "Number too large to set time in future " + str(hours_in_future)
+
+def get_datetime(currtime: datetime, hours: int):
+    """
+    Gets the time n hours in the future or past
+
+    Parameters:
+    currtime (datetime): A datetime object with the starting datetime.
+    hours (int): Hours going forward or backwards
+
+    Returns:
+    datetime: A datetime object
+    """
+    # TODO: input sanitation
+    # Calculate lower boundary, hours from year 1 to currtime:
+    timedelta_lowerbound = currtime - datetime(1, 1, 1)  # 1st Jan 0001
+    hours_min = timedelta_lowerbound.total_seconds() // 3600 * -1
+
+    # Calculate upper boundary, hours from year 9999 to currtime:
+    timedelta_upperbound = datetime(9999, 12, 31) - currtime
+    hours_max = timedelta_upperbound.total_seconds() // 3600
+
+    # TODO: proper secure logging
+    logging.debug(
+        "hours_max=%s hours_min=%s hours=%s",
+        hours_max,
+        hours_min,
+        hours,
+    )
+    if (hours > hours_max) or (hours < hours_min):
+        raise ValueError("hours out of range")
+
+    return currtime + timedelta(hours=hours)
 
 
 #####################
 # attempting to exploit above code example
 #####################
-print(get_time_in_future(23**74))
+datetime.fromtimestamp(0)
+currtime = datetime.fromtimestamp(1)  # 1st Jan 1970
```

Comment:
Maybe we should change the name of the "currtime" variable to be different than the input parameter into the "get_datetime()" function. It's better practice

```suggestion
current_time = datetime.fromtimestamp(1)  # 1st Jan 1970
```